feat: Add state root calculation time from latest flashblock state to metering RPC

niran opened the pull request:
Adds state root calculation timing with flashblock state support to the metering RPC to provide visibility into state root computation performance as an independent metric.

### TODO

- [X] Add tests with state changes across blocks and flashblocks based on the upcoming refactored flashblocks-rpc test suite

### Key Changes

**1. Flashblock State Integration**
- Bundles now execute on top of pending flashblock state (not just canonical block)
- Accurately accounts for nonces, balances, storage, and code changes from pending transactions
- Adds `state_flashblock_index` to response

**2. State Root Optimization**
- New `FlashblockTrieCache`: caches latest flashblock's trie nodes
- Computes flashblock state root once, reuses for all subsequent bundle metering calls
- Cache keyed by `(block_hash, flashblock_index)`
- Uses `TrieInput.prepend_cached()` to combine cached flashblock trie with bundle changes
- Significant performance improvement for repeated metering on same flashblock

**3. State Root Metrics**
- Separately tracks state root calculation time
- Flashblock trie computation excluded from timing metrics

**4. Flashblocks-RPC Fixes (discovered during implementation)**
- **Fixed incorrect database layering:** Changed from `CacheDB → State` to `State → CacheDB → StateProviderDatabase`
- **Added bundle state tracking:** `PendingBlocks` now tracks accumulated `BundleState` alongside cache
- **Added state accumulation:** Uses `with_bundle_prestate()` to accumulate state changes across flashblocks
- These fixes enable accurate state root calculations for metering and potentially benefit other flashblock state consumers

niran committed:
Include state root calculation time in metering

Add state root calculation after transaction execution and measure its
time independently. The state root calculation time is included in the
total execution time and also tracked separately.

niran committed:
Update tips-core and expose state root time in RPC response

Update tips-core dependency to include state_root_time_us field in
MeterBundleResponse. This exposes the state root calculation time as
an independent metric in the metering RPC response.

niran committed:
Add tests for state root time metrics

Add unit test to verify total_execution_time_us >= state_root_time_us
invariant and RPC integration test to verify state_root_time_us is
properly exposed in the response.

niran committed:
Simplify state root calculation by removing unnecessary ExecutionOutcome wrapper

niran committed:
Use pending flashblocks state for bundle metering

Integrate flashblocks state into metering to execute bundles on top of
pending flashblock state rather than canonical block state. This ensures
metered gas usage and execution time accurately reflect the effects of
pending transactions (nonces, balances, storage, code changes).

Implementation:
- Add flashblocks-rpc dependency to metering crate
- Update meter_bundle() to accept optional db_cache parameter
- Implement three-layer state architecture:
  1. StateProviderDatabase (canonical block base state)
  2. CacheDB (applies flashblock pending changes via cache)
  3. State wrapper (for EVM builder compatibility)
- Update MeteringApiImpl to accept FlashblocksState
- Get pending blocks and db_cache from flashblocks when available
- Fall back to canonical block state when no flashblocks available
- Update response to include flashblock_index in logs
- Require flashblocks to be enabled for metering RPC
- Update all tests to pass FlashblocksState parameter

The metering RPC now uses the same state as flashblocks eth_call,
ensuring consistent simulation results.

niran committed:
Cache flashblock trie nodes to optimize bundle metering

Adds a single-entry cache for the latest flashblock's trie nodes, allowing
bundle metering operations to reuse the cached trie computation instead of
recalculating from scratch each time.

Key changes:
- Add FlashblockTrieCache: thread-safe single-entry cache for latest flashblock
- Add FlashblockTrieData: contains trie updates and hashed state for reuse
- Cache keyed by block hash and flashblock index for accuracy
- Compute flashblock trie once per flashblock, reuse for all bundle operations
- Extract flashblock trie calculation outside timing metrics
- Use TrieInput.prepend_cached() to combine flashblock + bundle tries

The cache replaces previous entries when a new flashblock is cached, as it's
designed only for the current/latest flashblock, not historical ones.

niran committed:
linter fixes

niran committed:
Rename total_execution_time to total_time for clarity

The timer and field names now better reflect that they measure the total
elapsed time (including both transaction execution and state root
calculation), rather than just execution time. This distinguishes them
from the per-transaction execution_time_us metrics.

Changes:
- Renamed execution_start → total_start timer
- Renamed total_execution_time → total_time variable
- Renamed total_execution_time_us → total_time_us in MeterBundleOutput
- Updated logging and tests to use new naming
- Added TODO comment for tips-core field rename

niran committed:
Refactor flashblock trie data handling to reduce code duplication

Consolidate the logic for using cached vs computed flashblock trie data by introducing a single flashblock_trie_data variable that handles both cases. This eliminates duplicate code paths in the state root calculation section.

niran committed:
Add chain-state metering tests and flashblock harness support

niran committed:
Fix chain-state metering tests with L1 deposit

niran committed:
Restore beacon root for successful metering tests

niran committed:
Have storage persistence test meter the next block

niran committed:
Rename metering tests for clarity

niran committed:
Document why flashblock trie caching isolates bundle I/O

Add comments explaining that the trie cache ensures each bundle's
state root calculation measures only the bundle's incremental I/O,
not the accumulated I/O from previous flashblocks.

niran committed:
Move build_single_flashblock from metering to test-utils

niran committed:
tweak comments

niran committed:
Update tips-core and fix metering harness/tests

niran committed:
Silence the global executor errors

niran committed:
Add TODO for state_root_time

danyalprout commented on crates/metering/Cargo.toml:
```diff
@@ -26,17 +26,25 @@ reth-optimism-chainspec.workspace = true
+op-alloy-rpc-types.workspace = true
+
+# base
+base-reth-flashblocks-rpc = { path = "../flashblocks-rpc" }
```
Can we use a workspace setting for this?

niran committed:
Add metrics for bundle state clone cost

Track duration and size (number of accounts) when cloning BundleState
to help identify if this becomes a performance bottleneck.

niran commented:
> > Fixed incorrect database layering: Changed from CacheDB → State to State → CacheDB → StateProviderDatabase
> 
> What issues existed previously and what implications does this have?

`CacheDB` is a read-through cache. Writes are not propagated. As a result, I believe the `State` layer in the previous implementation wasn't doing anything. `State` is for producing `BundleStates` that the rest of the reth machinery can use to persist the changes, or in our case, calculate the state root. If a `CacheDB` is in front of it, no changes will be written to the `State`.

We can either remove the `CacheDB` entirely, or we can keep it around like the current PR does to maintain a cache of reads.

I could be wrong about this since I haven't actually written code that compares both outcomes. But that's my reading of the docs.

> Definitely think this needs to go on mainnet dev node before being merged

Makes sense.

niran committed:
Add database layering tests to flashblocks-rpc

Add tests verifying the correct State<CacheDB<StateProviderDatabase>>
layering order. These tests demonstrate that CacheDB<State<...>> loses
writes because CacheDB intercepts all writes into its internal cache
without propagating them to State's bundle tracking.

niran commented on crates/flashblocks-rpc/src/pending_blocks.rs:
```diff
@@ -195,6 +208,10 @@ impl PendingBlocks {
 
+    pub fn get_bundle_state(&self) -> BundleState {
+        self.bundle_state.clone()
+    }
```
Added metrics for this in 9752d0ab008d02fa996e82dc12e127dc44d9c89e.

niran committed:
Add tests for flashblock state visibility

Add two tests that verify pending state from flashblocks is correctly
visible:

- test_eth_call_sees_flashblock_state_changes: Verifies eth_call to
  pending block sees balance changes from executed flashblock
  transactions

- test_sequential_nonces_across_flashblocks: Verifies transactions in
  flashblock N+1 can see state changes (nonces) from flashblock N

niran committed:
just fix

niran committed:
Fix build issues after rebase

- Add missing BundleRetention import in flashblocks processor
- Add PendingBlocksAPI import in meter_rpc
- Export FlashblockTrieCache, FlashblockTrieData, FlashblocksState from rpc crate
- Update runner to require flashblocks for metering API
- Restore test-utils files from main

niran committed:
Fix clippy warnings and apply nightly fmt

- Add missing doc comments for bundle state clone metrics
- Remove unused crate dependencies (alloy-sol-types, op-alloy-consensus, revm)
- Fix collapsible_if lint in flashblock_trie_cache.rs
- Add Debug derive to MeteringApiImpl
- Export MeterBundleOutput from public API

niran committed:
Fix build issues after rebase

- Fix load_account -> load_cache_account in processor.rs (State API differs from CacheDB)
- Add FB generic parameter to MeteringApiImpl for flashblocks support
- Fix revm module paths in tests (revm:: -> reth::revm::)
- Add DummyFlashblocks implementation for meter_rpc tests
- Add dev-dependencies: alloy-genesis, reth-optimism-node
- Remove non-existent pub mod flashblocks from test-utils

danyalprout commented:
Closing temp until after rebase
